### PR TITLE
Refactor path reconstruction for shortest path

### PR DESCRIPTION
**Pull request by joweich**

### CODE REVIEW

File: include/graaflib/algorithm/shortest_path.tpp [hunk 1]
```diff
@@ -31,53 +50,29 @@ std::optional<GraphPath<WEIGHT_T>> get_unweighted_shortest_path(
...
-        return v1.distance > v2.distance;
-      });
+  std::unordered_map<vertex_id_t, PathVertex<WEIGHT_T>> vertex_info;
+  std::priority_queue<PathVertex<WEIGHT_T>, std::vector<PathVertex<WEIGHT_T>>,
+                      std::function<bool(const PathVertex<WEIGHT_T>&,
+                                         const PathVertex<WEIGHT_T>&)>>
+      to_explore(
+          [](const PathVertex<WEIGHT_T>& v1, const PathVertex<WEIGHT_T>& v2) {
+            return v1.dist_from_start > v2.dist_from_start;
+          });
```
Comment (bobluppes) [inline include/graaflib/algorithm/shortest_path.tpp:75]: What do you think about making this comparison part of the `PathVertex` type? 

```c++
template <typename E>
struct PathVertex {
  // data members

  [[nodiscard]] bool operator>(const PathVertex<E>& other) {
    return dist_from_start > other.dist_from_start;
  }
};
```

Then the type of this priority queue can be simplified to
```suggestion
  using weighted_path_item = PathVertex<WEIGHT_T>;
  std::priority_queue<weighted_path_item, std::vector<weighted_path_item>, std::greater<weighted_path_item>> to_explore{};
```

**Review by bobluppes — APPROVED** [review]
Nice refactor, I like 👍🏻 
Just some smaller ideas for further improvement

File: include/graaflib/algorithm/shortest_path.h [hunk 2]
```diff
@@ -22,6 +22,13 @@ struct GraphPath {
   }
 };
 
+template <typename E>
+struct PathVertex {
+  vertex_id_t id;
+  E dist_from_start;
+  vertex_id_t prev_id;
+};
+
```
[the remark this answers is not in the document]
Comment (joweich) [reply include/graaflib/algorithm/shortest_path.h]: Good chatches!

File: include/graaflib/algorithm/shortest_path.tpp (hunk 1 above)
Comment (joweich) [reply include/graaflib/algorithm/shortest_path.tpp]: Done! I think you don't need the `weighted_path_item` in `std::greater` though

**Comment by bobluppes** [comment]
Very nice, LGTM 🚀

**Pull request merged**